Add frontend to taxi-app

Sir-Beard opened the pull request:
Adding jsp, css, controllers(servlets) to the project code.

Sir-Beard committed:
Taxi service. Task4. Commit 1.
Logger. Adding log4j2 to the project.
Updated dao classes with logger.
timestamp: 23:02 09.07.2023

Sir-Beard committed:
Merge branch 'add-car-service-functionality' into add-logger-log4j2

# Conflicts:
#	src/main/java/com/taxi/Main.java
#	src/main/java/com/taxi/dao/CarDaoImpl.java
#	src/main/java/com/taxi/dao/DriverDaoImpl.java
#	src/main/java/com/taxi/dao/ManufacturerDaoImpl.java

timestamp: 01:46 19.07.2023

Sir-Beard committed:
Merge branch 'add-car-service-functionality' into add-spock-framework

Sir-Beard committed:
Taxi service. Task5. Commit 1.
Add SpockFramework + Groovy tests. Test commit.
timestamp: 21:56 19.07.2023

Sir-Beard committed:
Merge branch 'add-car-service-functionality' into add-spock-framework

# Conflicts:
#	.gitignore

Sir-Beard committed:
Taxi service. Task 5. Commit 1. TEST commit.
just testing commit of some test files + task file of Spock+groovy tests.
timestamp: 01:12 28.07.2023

Sir-Beard committed:
Taxi service. Task 6. Commit 1. Push request 1.
Adding new code functionality to the application:
jsp(+jstl, css), java controllers
timestamp: 19:05 25.08.2023

Rommelua commented on src/main/java/com/taxi/controller/AddDriverToCarController.java:
```diff
@@ -0,0 +1,47 @@
+import javax.servlet.http.HttpServletResponse;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+
+@WebServlet("/AddDriverToCarController")
```
https://mate-academy.github.io/jv-program-common-mistakes/java-web/web-practice/java-web-practice

Sir-Beard replied:
Виправив. 
Використав один стиль для неймінгу URLs як вказано в common mistakes.
Переписав анотацію мапінгу webservlet в контроллерах і лінки в jsp сторінках.

Rommelua commented on src/main/java/com/taxi/controller/AddDriverToCarController.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        getServletContext().getRequestDispatcher("/WEB-INF/views/addDriverToCar.jsp").forward(req, resp);
```
```suggestion
        req.getRequestDispatcher("/WEB-INF/views/addDriverToCar.jsp").forward(req, resp);
```

Sir-Beard replied:
Виправив.

Rommelua commented on src/main/java/com/taxi/controller/DisplayAllDriversController.java:
```diff
@@ -0,0 +1,50 @@
+        req.getRequestDispatcher("/WEB-INF/views/displayAllDrivers.jsp").forward(req, resp);
+    }
+
+    private Set<Driver> sortById(Set<Driver> allDrivers) {
+        List<Driver> list = new ArrayList<>(allDrivers);
+        boolean isSorted = false;
+        while (!isSorted) {
+            isSorted = true;
+            for (int i = 0; i < list.size() - 1; i++) {
+                if (list.get(i).getId() > list.get(i + 1).getId()) {
+                    list.add(i, list.get(i + 1));
+                    list.remove(i + 2);
+                    isSorted = false;
+                }
+            }
+        }
+        allDrivers = new LinkedHashSet<>(list);
+        return allDrivers;
```
Не надо так делать

Sir-Beard replied:
Видалив цей смішний непотрібний метод.
Зробив красиво: в класі DriverDaoImpl.java в методі getAll() замінив HashSet на TreeSet, і в класі Driver.java імплементував інтерфейс Comparable з імплементацією метода compareTo(Driver o).

Rommelua commented on src/main/java/com/taxi/controller/DriverController.java:
```diff
@@ -0,0 +1,37 @@
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+
+@WebServlet("/DriverController")
+public class DriverController extends HttpServlet {
```
```suggestion
public class AddDriverController extends HttpServlet {
```

Sir-Beard replied:
змінив назви на більш системні та зрозумілі там, де логіка говорить про **створення** моделі

Rommelua commented on src/main/java/com/taxi/controller/DriverDeleteController.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+
+@WebServlet("/DriverDeleteController/delete")
+public class DriverDeleteController extends HttpServlet {
```
```suggestion
public class DeleteDriverController extends HttpServlet {
```

Sir-Beard replied:
змінив назви на більш системні та зрозумілі там, де логіка говорить про **видалення** моделі

Rommelua commented on src/main/java/com/taxi/controller/ManufacturerController.java:
```diff
@@ -0,0 +1,37 @@
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+
+@WebServlet("/ManufacturerController")
+public class ManufacturerController extends HttpServlet {
```
```suggestion
public class AddManufacturerController extends HttpServlet {
```

Sir-Beard replied:
Виправив.

Sir-Beard committed:
Taxi service. Task 6. Commit 2.
1.Renaming the names of classes (controllers) to be more systematicity/logic.
2.Used one style for naming URLs. Rewrote the webservlet mapping annotations in controllers and links in jsp pages.
3.Instead of the ridiculous sortById method in DisplayAllDriversController.java, I implemented a comparator in the Driver.java class with an implementation of the compareTo method.
+ In DriverDaoImpl.java replaced HashSet with TreeSet implementation.
4.Minor corrections.
timestamp: 00:56 30.08.2023

Rommelua commented on src/main/java/com/taxi/controller/AddCarController.java:
```diff
@@ -0,0 +1,43 @@
+        car.setManufacturer(manufacturer);
+        car.setModel(model);
+        carService.create(car);
+        resp.sendRedirect(req.getContextPath() + "/");
```
Ты всегда в ПОСТ методах редиректишь на стартовую страницу. Лучше делать немного по-другому. Например тут более логично направить на get all cars

Sir-Beard replied:
Переробив всі сервлети, додав логічні переходи на сторінки.

Rommelua commented on src/main/java/com/taxi/controller/DeleteManufacturerController.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+
+@WebServlet("/manufacturers/delete")
+public class DeleteManufacturerController extends HttpServlet {
+    private final AnnotationConfigApplicationContext applicationContext
+            = new AnnotationConfigApplicationContext(AppConfig.class);
```
Кстати. Можно сделать AbstractController extends HttpServlet, положить в него AnnotationConfigApplicationContext, и наследовать от него все сервлеты.

Sir-Beard replied:
Створив AbstractController класс який наслідує HttpServlet. І переробив всі свої сервлети наслідуючи AbstractController. Гарна ідея, спрацювала, дякую :)

Rommelua commented on src/test/groovy/ManufacturerServiceSpec.groovy:
```diff
@@ -0,0 +1,33 @@
+    def "should create Manufacturer with name and country"() {
+        given:
+        ManufacturerService manufacturerService = Mock(ManufacturerService)
+        def manufacturer = new Manufacturer(0, "Dodge", "Canada", false)
+        manufacturerService.create(manufacturer) >> manufacturer
+
+        when:
+        def result = manufacturerService.create(manufacturer)
+
+        then:
+        result.id == 0
+        result.name == "Dodge"
+        result.country == "Canada"
```
Тут нет никакого тестирования сервиса)))

Sir-Beard replied:
Та, ще розбираюсь

Rommelua reviewed (approved):
Approve apart from tests

Sir-Beard committed:
Taxi service. Task 6. Commit 3.
1. Adding one AbstractController.java class to hold HttpServlet.
2. Rewritten Controller(Servlet) classes to inheritance main AbstractController(Servler) class.
3. Rewritten controller classes to for logical redirection to relevant jsp pages.
timestamp: 15:44 03.09.2023

Sir-Beard committed:
Taxi service. Task 6. Commit 4.
Maven checkout.
timestamp: 15:45 03.09.2023
